source/slang/slang-check-decl.cpp

```diff
@@ -3220,6 +3225,132 @@ void SemanticsDeclHeaderVisitor::validateGenericConstraintSubType(
+    // If that's not the case, then early out.
+    auto parentGeneric = as<GenericDecl>(decl->parentDecl);
+    if (!parentGeneric)
+        return;
+
+    // Build a HashSet of all generic parameter declarations seen before the constraint.
+    // After we collect the referenced declarations from the constraint's superior type,
+    // we can do a quick check with the HashSet to see if a reference falls outside of
+    // the set, indicating that we have a forward reference.
+    HashSet<Decl*> declaredBeforeConstraint;
+    bool foundConstraint = false;
+
+    for (Index i = 0; i < parentGeneric->getDirectMemberDeclCount(); ++i)
+    {
+        auto member = parentGeneric->getDirectMemberDecl(i);
+
+        if (member == decl)
+        {
+            foundConstraint = true;
+            break;
+        }
+
+        // Add generic type parameters to our "declared so far" set
+        if (auto typeParam = as<GenericTypeParamDeclBase>(member))
+        {
+            declaredBeforeConstraint.add(typeParam);
+        }
+    }
+
+    // This probably shouldn't happen, but if the constraint is not found in parent GenericDecl,
+    // just early out as we can't check forward references.
+    if (!foundConstraint)
+        return;
+
+    // Collect all referenced declarations from the constraint's superior type
+    HashSet<Decl*> referencedDecls;
+    collectReferencedDecls(decl->sup.type, referencedDecls);
+
+    // Check if any of the referenced declarations are forward references (not in our "declared so
+    // far" set)
+    for (auto referencedDecl : referencedDecls)
+    {
+        if (auto typeParam = as<GenericTypeParamDeclBase>(referencedDecl))
+        {
+            // Check if this type parameter belongs to the same generic but is NOT in our "declared
+            // so far" set
+            if (typeParam->parentDecl == parentGeneric &&
+                !declaredBeforeConstraint.contains(typeParam))
+            {
+                // Found a forward reference, report an error.
+
+                // Try to get names of the forward reference and
+                // the constraint parameter for better error info.
+                Name* forwardParamName = typeParam->getName();
+
+                Name* constraintParamName = nullptr;
+                if (auto subDeclRefType = as<DeclRefType>(decl->sub.type))
+                {
+                    if (auto subDecl = subDeclRefType->getDeclRef().getDecl())
+                    {
+                        constraintParamName = subDecl->getName();
+                    }
+                }
+
+                // Report the error using available names
+                if (constraintParamName && forwardParamName)
+                {
+                    getSink()->diagnose(
+                        decl->sup.exp,
+                        Diagnostics::forwardReferenceInGenericConstraint,
+                        constraintParamName,
+                        forwardParamName);
```

Comment:
You can just pass in `typeParam` and `decl->sub.type` directly, no need to gather the name first.